snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/request/SnomedQueryOptimizer.java

```diff
@@ -15,169 +15,847 @@
+		hierarchyStats.filterRedundantNoFalsePositives(conceptSet);
+	}
+
+	private void filterAncestorsForExclusion(final SnomedHierarchyStats hierarchyStats, final Set<String> inclusionAncestors) { 
+		// No ancestor of an included concept can be used for exclusions -- it would remove at least one concept we want to keep in the set. 
+		hierarchyStats.removeCandidates(inclusionAncestors);
+
+		// a) ancestor does not cover the minimum required members of the set
+		hierarchyStats.filterByClusterSize(minimumClusterSize);
+		// b) a non-member with a single member as its child -- there is no reason to replace "=child" with "<parent" in this case
+		hierarchyStats.filterSingleChildMember();
+		// c) more false positive children than true positive children
+		hierarchyStats.filterByChildren();
+		// d) ancestor has no false positives, but it has an _ancestor_ which has no false positives either (or at most 1 and is a non-member)
+		hierarchyStats.filterRedundantNoFalsePositives(conceptsToExclude);
+	}	
+
+	// Compute scores, then iterate over the best candidates with an acceptable fit in decreasing order
+	private void optimizeAncestorInclusions(final BranchContext context, final SnomedHierarchyStats inclusionHierarchyStats) {
+
+		final LongKeyFloatMap ancestorScores = inclusionHierarchyStats.initAncestorScores(
+			context, 
+			conceptSet, 
+			optimizerStrategy, 
+			zoom, 
+			weight);
+
+		if (ancestorScores.isEmpty()) {
+			return;
+		}
+		
+		final Instant startTime = Instant.now(clock);
+		final Instant endTime = startTime.plus(maxRuntime);
+
+		int iteration = 1;
+		int ancestorExpressionCount = 0;
+		boolean ancestorFound = false;
+		boolean canceled = false;
+
+		while (!canceled) {
+			final String bestFitAncestor = getBestAncestor(context, ancestorScores);
+
+			if (bestFitAncestor != null) {
+				ancestorFound = true;
+				ancestorExpressionCount++;
+
+				// Convert ancestor to an inclusion expression
+				final boolean member = conceptSet.contains(bestFitAncestor);
+				final String operator = member ? "<<" : "<";
+				final QueryExpression bestFitInclusion = new QueryExpression(IDs.base62UUID(), String.format("%s %s", operator, bestFitAncestor), false);
+
+				// Add inclusion to our optimized list
+				optimizedInclusions.add(bestFitInclusion);
+
+				// "applyInclusions" is effectively inlined here because we need access to the number of concepts included
+				final Set<String> includedIds = evaluateEclToSet(context, bestFitInclusion.getQuery());
+				final Set<String> expected = Sets.intersection(includedIds, conceptSet);
+				conceptsToInclude.removeAll(expected);
+
+				final Set<String> unexpected = Sets.difference(includedIds, conceptSet);
+				conceptsToExclude.addAll(unexpected);
+
+				// Tune parameters after a successful selection
+				if (((float) conceptSet.size()) / (includedIds.size() + 1) > 1000.0f) {
```

Comment:
What is 1000 here? 😕

Reply:
If every clause would cover as many concepts from the evaluated concept set as the one that we have just found, we'd have about 1000 clauses. This is an indication that the iteration progresses too slowly.